### PR TITLE
Update integration tests to use latest Chrome, Firefox and Chrome Dev

### LINKED ISSUE #18784: Run tests in Chrome Dev as well.
We should run our unit and functional tests on Chrome Canary in addition to latest Chrome. 

This will help catch issues such as #18777 a few weeks earlier.

/to @rsimha

### PR DESCRIPTION
**Pull request by estherkim**
Fixes #18784 

Updates Sauce Labs platforms to use latest Chrome and Firefox versions automatically.

### CODE REVIEW

File: build-system/tasks/runtime-test.js [hunk 1]
```diff
@@ -73,21 +73,20 @@ function getConfig() {
...
-        // 'SL_Edge_17',
-        // 'SL_IE_11',
+        'SL_Chrome',
+        'SL_Chrome_Dev',
+        'SL_Firefox',
+        'SL_Safari_12',
+        'SL_Chrome_Android_7',
+        'SL_iOS_12',
+        'SL_Edge_17',
+        'SL_IE_11',
```
Comment (rsimha) [inline build-system/tasks/runtime-test.js]: I suspect the proxy will get overloaded if we add so many platforms in one go. An alternative is what we talked about yesterday: split into two sets and run one after another as two consecutive integration test invocations.

Comment (rsimha) [reply build-system/tasks/runtime-test.js]: IIRC, `concurrency` only controls the number of platforms on which tests run in parallel. It doesn't restrict the number of tunnels created. Worth experimenting with, though.

Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Yup you're right. Batched the karma test instead with `const batchSize = 4;` which seems to work okay with the SauceLabs tunnels.

File: build-system/tasks/runtime-test.js [hunk 2]
```diff
@@ -541,96 +546,123 @@ function runTests() {
...
-      console./* OK*/log('travis_fold:end:console_errors_' + sectionMarker);
-    }
-    server.emit('kill');
-    if (exitCode) {
-      log(
-          red('ERROR:'),
-          yellow('Karma test failed with exit code ' + exitCode));
+  if (argv.saucelabs || argv.saucelabs_lite) {
+    let batch = 1;
+    const batchSize = 4; // Number of Sauce Lab browsers
```
Comment (danielrozenberg) [inline build-system/tasks/runtime-test.js]: Make this a top-level constant

File: build-system/tasks/runtime-test.js [hunk 3]
```diff
@@ -541,96 +546,123 @@ function runTests() {
...
-    if (shouldCollapseSummary) {
-      console./* OK*/log('travis_fold:end:console_errors_' + sectionMarker);
-    }
-    server.emit('kill');
-    if (exitCode) {
-      log(
-          red('ERROR:'),
-          yellow('Karma test failed with exit code ' + exitCode));
+  if (argv.saucelabs || argv.saucelabs_lite) {
+    let batch = 1;
```
Comment (danielrozenberg) [inline build-system/tasks/runtime-test.js]: I don't understand what this block does. Care to add a comment explaining the algorithm?

File: build-system/tasks/runtime-test.js [hunk 4]
```diff
@@ -541,96 +546,123 @@ function runTests() {
...
+    const batchSize = 4; // Number of Sauce Lab browsers
+    let left = 0;
+    let right = batchSize;
+    while (left < right) {
+      const configBatch = c;
```
Comment (danielrozenberg) [inline build-system/tasks/runtime-test.js]: What does this line do?

File: build-system/tasks/visual-diff/index.js [hunk 5]
```diff
@@ -630,7 +630,8 @@ async function ensureOrBuildAmpRuntimeInTestMode_() {
 
 function installPercy_() {
   log('info', 'Running', colors.cyan('yarn'), 'to install Percy...');
-  execOrDie('npx yarn --cwd build-system/tasks/visual-diff');
+  execOrDie('npx yarn --cwd build-system/tasks/visual-diff',
+      {'stdio': 'ignore'});
```
Comment (danielrozenberg) [inline build-system/tasks/visual-diff/index.js]: nit: prefer splitting minor unrelated edits to separate PRs - it's totally fine to create a PR with 1 line! (but since you've already done this, leave this in)

File: build-system/tasks/runtime-test.js (hunk 4 above)
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Loops through the browser list in batches. I changed "left" and "right" to "startIndex" and "endIndex" for clarity.

File: build-system/tasks/runtime-test.js (hunk 3 above)
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Done!

File: build-system/tasks/runtime-test.js (hunk 2 above)
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Done!

File: build-system/tasks/runtime-test.js (hunk 4 above)
Comment (danielrozenberg) [reply build-system/tasks/runtime-test.js]: Creating a new `const` here might not do what you think it does :)

The "constant-ness" of `configBatch` only means that you cannot change what object is being pointed to by `configBatch` (e.g., you can't do `configBatch = null` afterwards inside this block). It doesn't create a copy - so any changes to `configBatch` will also change `c`:

```javascript
~$ node
> const a = {1: 'hello', 2: 'world'}
undefined
> const b = a;
undefined
> b[2] = 'kitty!'
'kitty!'
> b
{ '1': 'hello', '2': 'kitty!' }
> a
{ '1': 'hello', '2': 'kitty!' }
> 
```

Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Makes sense :) Updated to copy the config instead

File: extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js [hunk 6]
```diff
@@ -62,15 +62,15 @@ describe.configure().run('amp-fx-collection', function() {
...
+    it.skip('runs fly-in-left animation with default parameters', () => {
       expect(isExperimentOn(win, 'amp-fx-fly-in')).to.be.true;
       const initialLeft = getComputedLeft(win);
       win.scrollTo(0, 0.5 * getViewportHeight(win));
       return timeout(2000).then(() => {
         expect(getComputedLeft(win)).to.be.above(initialLeft);
       });
     });
-  });
+  }); //TODO: fails in Firefox 63.0.0
```
Comment (rsimha) [inline extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js]: For style consistency, move the `TODO` to the line above the `it.skip` and annotate it similar to this example:

https://github.com/ampproject/amphtml/blob/bf11b71ee0154df6689ab191f6a78785927861ea/test/integration/test-video-players-helper.js#L92-L93

File: extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js [hunk 7]
```diff
@@ -97,15 +97,15 @@ describe.configure().run('amp-fx-collection', function() {
...
+    it.skip('runs fly-in-right animation with default parameters', () => {
       expect(isExperimentOn(win, 'amp-fx-fly-in')).to.be.true;
       const initialLeft = getComputedLeft(win);
       win.scrollTo(0, 0.5 * getViewportHeight(win));
       return timeout(2000).then(() => {
         expect(getComputedLeft(win)).to.be.below(initialLeft);
       });
     });
-  });
+  }); //TODO: fails in Firefox 63.0.0
```
Comment (rsimha) [inline extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js]: Here too.

File: build-system/tasks/karma.conf.js [hunk 8]
```diff
@@ -132,7 +132,7 @@ module.exports = {
   ],
 
   // Number of sauce platforms to start in parallel
-  concurrency: 4,
+  concurrency: Infinity,
```
Comment (rsimha) [inline build-system/tasks/karma.conf.js]: From the Karma [docs](http://karma-runner.github.io/3.0/config/configuration-file.html), `Infinity` is the default concurrency, so you can just remove this config setting.

File: build-system/tasks/runtime-test.js [hunk 9]
```diff
@@ -541,96 +547,124 @@ function runTests() {
...
+    while (startIndex < endIndex) {
+      const configBatch = Object.assign({}, c);
+      configBatch.browsers = saucelabsBrowsers.slice(startIndex, endIndex);
+      log(green('Batch #' + batch + ': Running tests on ' +
+        configBatch.browsers.length + ' Sauce Labs browser(s)...'));
+      await createKarmaServer(configBatch);
+      startIndex = batch * batchSize;
+      batch++;
+      endIndex = Math.min(batch * batchSize, saucelabsBrowsers.length);
     }
```
Comment (rsimha) [inline build-system/tasks/runtime-test.js]: For readability, refactor this entire batch-running code block into a separate function.

File: build-system/tasks/runtime-test.js [hunk 10]
```diff
@@ -541,96 +547,124 @@ function runTests() {
...
+  // Sauce Labs environment.
+  if (argv.saucelabs || argv.saucelabs_lite) {
+    let batch = 1;
+    let startIndex = 0;
+    let endIndex = batchSize;
+    while (startIndex < endIndex) {
+      const configBatch = Object.assign({}, c);
+      configBatch.browsers = saucelabsBrowsers.slice(startIndex, endIndex);
+      log(green('Batch #' + batch + ': Running tests on ' +
+        configBatch.browsers.length + ' Sauce Labs browser(s)...'));
```
Comment (rsimha) [inline build-system/tasks/runtime-test.js]: Add a log statement outside this loop indicating the total number browsers on which tests are being run.

File: build-system/tasks/runtime-test.js [hunk 11]
```diff
@@ -541,96 +547,124 @@ function runTests() {
...
+    await createKarmaServer(c);
+  }
+  server.emit('kill');
+  exitCtrlcHandler(handlerProcess);
+
+  /**
+   * Creates and starts karma server
+   *
+   * @param {!Object} configBatch
+   *
```
Comment (rsimha) [inline build-system/tasks/runtime-test.js]: For conciseness, delete blank lines in JSDoc. (I believe VSCode will automatically format JSDoc for you if you start off typing `/**` on a new line.)

File: extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js (hunk 7 above)
Comment (estherkim) [reply extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js]: Done.

File: extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js (hunk 6 above)
Comment (estherkim) [reply extensions/amp-fx-collection/0.1/test/integration/test-amp-fx-fly-in.js]: Done.

File: build-system/tasks/karma.conf.js (hunk 8 above)
Comment (estherkim) [reply build-system/tasks/karma.conf.js]: Removed.

File: build-system/tasks/runtime-test.js (hunk 10 above)
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Done.

File: build-system/tasks/runtime-test.js (hunk 9 above)
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Done.

File: build-system/tasks/runtime-test.js (hunk 11 above)
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Removed.

File: build-system/tasks/runtime-test.js [hunk 12]
```diff
@@ -541,96 +547,124 @@ function runTests() {
...
-        process.exit(1);
+      // TODO(rsimha, 14814): Remove after Karma / Sauce ticket is resolved.
+      // if (process.env.TRAVIS) {
+      //   setTimeout(() => {
+      //     log(green('Process timed out'));
+      //     process.exit(exitCode);
+      //   }, 5000);
+      // } else {
+      //   process.exitCode = exitCode;
+      // }
```
[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: Done! I set the `exitCode` elsewhere to help with consolidating multiple exit codes. But the tests are now failing as expected.

**Comment by rsimha** [comment]
@estherkim Nice work so far! Let me know when this PR is ready for a final review, or if you need help in figuring out what to do with the failures.

File: build-system/tasks/runtime-test.js [hunk 13]
```diff
@@ -541,96 +549,141 @@ function runTests() {
...
+      ' Sauce Lab browsers in total'));
+    while (startIndex < endIndex) {
+      const configBatch = Object.assign({}, c);
+      configBatch.browsers = saucelabsBrowsers.slice(startIndex, endIndex);
+      log(green('Batch #' + batch + ': Running tests on ' +
+        configBatch.browsers.length + ' Sauce Labs browser(s)...'));
+      const batchExitCode = await createKarmaServer(configBatch);
+      if (batchExitCode && processExitCode == 0) {
+        processExitCode = batchExitCode;
+      }
```
Comment (rsimha) [inline build-system/tasks/runtime-test.js]: There might be a bug here, since I'm seeing that the second batch isn't running.

https://travis-ci.org/ampproject/amphtml/jobs/461286041#L3122

File: build-system/tasks/runtime-test.js [hunk 14]
```diff
@@ -541,96 +549,141 @@ function runTests() {
...
+   * @return {string} processExitCode
+   */
+  async function runTestInBatches() {
+    let batch = 1;
+    let startIndex = 0;
+    let endIndex = batchSize;
+    let processExitCode = 0;
+
+    log(green('Testing on ' + saucelabsBrowsers.length +
+      ' Sauce Lab browsers in total'));
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/tasks/runtime-test.js]: Also, you can use the backtick operator to avoid string concatenation:
```
log(green(`Testing on ${saucelabsBrowsers.length` Sauce Lab browser(s) in total`));
```

[the remark this answers is not in the document]
Comment (estherkim) [reply build-system/tasks/runtime-test.js]: @rsimha Updated the message.
@danielrozenberg Using backtip would be nice but it purposely includes line breaks which we don't want.

File: build-system/tasks/runtime-test.js [hunk 15]
```diff
@@ -541,96 +549,137 @@ function runTests() {
...
+   */
+  async function runTestInBatches() {
+    let batch = 1;
+    let startIndex = 0;
+    let endIndex = batchSize;
+    const batchExitCodes = [];
+
+    log(green('Running tests on ' + saucelabsBrowsers.length +
+      ' Sauce Lab browser(s) in total...'));
+    while (startIndex < endIndex) {
```
Comment (rsimha) [inline build-system/tasks/runtime-test.js]: Still seeing only one batch run here.

Comment (rsimha) [reply build-system/tasks/runtime-test.js]: Looks good this time!
